---
title: "Create data package"
author: "Trang Tran"
date: "March 22, 2020"
output: html_notebook
---

```{r setup, include=FALSE, echo=FALSE}
library(magrittr)
library(ggplot2)
library(Biobase)

options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
```

## Overview

## Download count matrices and meta data, set environment variable 'DBDIR'

```{r,eval = FALSE}
Sys.setenv('DBDIR' = '.')
```


## Gene-level `ExpressionSet`

### Assay data

```{r}
gene.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'matrix.gene.est_counts.RDS'))
gene.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'matrix.gene.tpm.RDS'))
gene.assayData <- new.env()
assign('exprs', gene.counts, envir = gene.assayData) # exprs is required
assign('count', gene.counts, envir = gene.assayData)
assign('tpm', gene.tpm, envir = gene.assayData)
ExpressionSet(assayData = gene.assayData)
```

### Phenotypic data

```{r}
metadata = read.table(file.path(Sys.getenv('DBDIR'),
                                'PRJNA282597_metadata_cleaned.tsv'), sep = '\t', header = TRUE)
libnames <- colnames(gene.assayData$exprs) %>% as.character()
phenoData <- data.frame('RunId' = libnames) %>%
    plyr::join(y = metadata, by = 'RunId', type = 'left')  %>%
    set_rownames(libnames) %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))   
```

### Annotations and features

```{r}
feature_attrs = read.table(file.path(Sys.getenv('DBDIR'),
                                     'feature_attributes.tsv'),
                           sep = '\t', header=TRUE, colClasses = 'character')
is.gene.NA = which(is.na(feature_attrs$gene_id))
feature_attrs[is.gene.NA, 'gene_id'] = feature_attrs[is.gene.NA, 'transcript_id']
gene.featureData <- data.frame('gene_id' = rownames(gene.assayData$exprs)) %>%
    plyr::join(y = feature_attrs[, c('gene_id', 'gene_biotype', 'gene_symbol')], by = 'gene_id', type = 'left') %>%
    set_names(c('ID', 'Biotype', 'Symbol')) %>%
    unique() 
rownames(gene.featureData) = gene.featureData$ID
gene.featureData = gene.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
```

### Experiment description

```{r}
experimentData <- new("MIAME",
                      lab = "Sarcoma Oncogenesis",
                      contact = "Frédéric Chibon",
                      title = "RNA-seq performed on sarcomas to identify various alterations",
                      abstract = "",
                      other = list(
                          citation = "Lesluyes T et al., Genomic and transcriptomic comparison of post-radiation versus sporadic sarcomas., Mod Pathol, 2019 Dec;32(12):1786-1794"
                      )
                      )
```

### Assembling an `ExpressionSet`

```{r}
sarcoma.rnaseq.gene <- ExpressionSet(assayData = gene.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = gene.featureData,
                             annotation = 'Homo_sapiens.GRCh38.99 + ERCC-92'
                             )
```

## Transcript-level `ExpressionSet`

### Assay data

```{r}
transcript.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'matrix.est_counts.RDS'))
transcript.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'matrix.tpm.RDS'))
transcript.assayData <- new.env()
assign('exprs', transcript.counts, envir = transcript.assayData) # exprs is required
assign('count', transcript.counts, envir = transcript.assayData)
assign('tpm', transcript.tpm, envir = transcript.assayData)
ExpressionSet(assayData = transcript.assayData)
```

### Phenotypic data

Same as above

### Annotations and features

```{r}
# feature_attrs = read.table(file.path(Sys.getenv('DBDIR'), 'feature_attributes.tsv'), sep = '\t', header=TRUE, colClasses = 'character')
transcript.featureData <- data.frame('transcript_id' = rownames(transcript.assayData$exprs)) %>%
    plyr::join(y = feature_attrs[, c('transcript_id', 'location', 'transcript_biotype')], by = 'transcript_id', type = 'left') %>%
    set_names(c('ID', 'Location', 'Biotype'))
rownames(transcript.featureData) = transcript.featureData$ID
transcript.featureData = transcript.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
```

### Experiment description

Same as above

### Assembling an `ExpressionSet`

```{r}
sarcoma.rnaseq.transcript <- ExpressionSet(assayData = transcript.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = transcript.featureData,
                             annotation = 'Homo_sapiens.GRCh38.99 + ERCC-92'
                             )
```

## Writing data sets

```{r,eval=FALSE}
usethis::use_data(sarcoma.rnaseq.gene,internal = FALSE, overwrite = TRUE)
usethis::use_data(sarcoma.rnaseq.transcript,internal = FALSE, overwrite = TRUE)
```

